### PR TITLE
Resource availability: backend changes

### PR DESCRIPTION
**Pull request by ChristopherChudzicki**
### What are the relevant tickets?
- https://github.com/mitodl/hq/issues/4743

### Description (What does it do?)
This PR adds a resource-level `availability: "dated" | "anytime" | null` field to the LearningResource model.

**ETL:** The availability data is calculated a bit differently per ETL source:
- `ocw, prolearn, micromasters, xpro` are constant values
    - Once https://github.com/mitodl/hq/issues/4953, xPRO could be read from API, but in practice, all are dated.
- `mitxonline` should read directly from its API.
    - set up to read from API, but for now returning `null` until https://github.com/mitodl/hq/issues/4881 is completed. *@JenniWhitman is working on it, at least for courses.*
- `mit_edx` and `oll` will be calculated based on edX API results
- videos, video playlists, podcasts, and podcast episodes have `availability: "anytime"`
- learning paths have `availability: null` 

**Frontend piece:** Plan to open a separate PR for the frontend piece—would like to get the backend portion reviewed. 

### How can this be tested?
1. Ensure you have all ETL settings from `settings_course_etl.py`. Then, check the ETL pipelines:
    ```
    ./manage.py backpopulate_ocw_data --skip-contentfiles
    ./manage.py backpopulate_mitxonline_data
    ./manage.py backpopulate_xpro_data
    ./manage.py backpopulate_prolearn_data
    ./manage.py backpopulate_oll_data
    # You may want to run this from a local datafile using
    # ./manage.py extract_openedx_data --etl_source mit_edx --out ./path/to/file
    # and adding --api_datafile ./path/to/file flag to command.
    ./manage.py backpopulate_mit_edx_data 
    ```
2. Now inspect the availability of various resources. I reocmmend a broad view via `dbshell` first: `./manage.py dbshell` then
    ```sql
    SELECT etl_source, array_agg(DISTINCT availability), etl_source, count(*)
    FROM learning_resources_learningresource
    WHERE published
    GROUP BY etl_source
    ORDER BY count(*) DESC;
    ```
3. You could also spot check some `mit_edx` resources since these are the ones that vary between dated / anytime right now. Still in `./manage.py dbshell`: 
    ```sql
    SELECT availability, url
    FROM learning_resources_learningresource
    WHERE etl_source = 'mit_edx' AND published
    ORDER BY availability;
    ```
    Open the URLs. If the resource is available anytime it should either be archived or self-paced and already started.

### Additional Context

Most programs still have `null` availability until we decide how to handle them.

### CODE REVIEW

**Review by mbertrand — COMMENTED** [review]
Looks great overall, just one minor suggestion, take or leave.

Also a question:
> ocw, prolearn, micromasters, xpro are constant values

micromasters always has null availability (that field hasn't been added to the ETL in this PR) - is its inclusion in the list above just a mistake?

**Review by mbertrand — APPROVED** [review]

File: learning_resources/etl/openedx.py [hunk 1]
```diff
@@ -136,6 +137,41 @@ def _get_course_marketing_url(config, course):
...
 
 
+def _get_run_published(course_run):
+    return course_run.get("status", "") == "published" and course_run.get(
+        "is_enrollable", False
+    )
+
+
+def _get_run_availability(course_run):
+    if course_run.get("availability") == "Archived":
```
[the remark this answers is not in the document]
Comment (ChristopherChudzicki) [reply learning_resources/etl/openedx.py:147]: Whoops. Yes, definitely.

**Comment by ChristopherChudzicki** [comment]
> > ocw, prolearn, micromasters, xpro are constant values
> 
> micromasters always has null availability (that field hasn't been added to the ETL in this PR) - is its inclusion in the list above just a mistake?

I waffled a bit a on this. It seems like we aren't entirely sure how Availability should behave for programs yet. (e.g., for mitxonline). I think I'll add dated for micromasters, though. It seems most appropriate.

**Review by mbertrand — APPROVED** [review]
👍

**Pull request merged**